Mathlib/LinearAlgebra/Matrix/Irreducible/Defs.lean

```diff
@@ -0,0 +1,253 @@
+      | @cons b _ q e =>
+        simp only [Quiver.Path.length_cons, Nat.succ.injEq] at hp_len
+        have h_Am_pos : 0 < (A ^ m) i b := (ih i b).mpr ⟨q, hp_len⟩
+        let h_A_pos := e
+        have h_prod : 0 < (A ^ m) i b * A b j := mul_pos h_Am_pos h_A_pos
+        exact
+          (Finset.sum_pos_iff_of_nonneg
+            (fun x _ => mul_nonneg (pow_apply_nonneg hA m i x) (hA x j))).2
+            ⟨b, Finset.mem_univ b, h_prod⟩
+
+/-- Irreducibility of a nonnegative matrix `A` is equivalent to entrywise positivity of some
+power: between any two indices `i, j` there exists a positive integer `k` such that the
+`(i, j)`-entry of `A ^ k` is strictly positive. -/
+theorem IsIrreducible.iff_exists_pow_pos
+    [IsOrderedRing R] [PosMulStrictMono R] [Nontrivial R] [DecidableEq n]
+    (hA : ∀ i j, 0 ≤ A i j) :
+    IsIrreducible A ↔ ∀ i j, ∃ k > 0, 0 < (A ^ k) i j := by
+  constructor
+  · intro h_irr i j
+    letI : Quiver n := toQuiver A
+    obtain ⟨p, hp_len⟩ := h_irr.2 i j
+    refine ⟨p.length, hp_len, ?_⟩
+    have : Nonempty {q : Path i j // q.length = p.length} := ⟨⟨p, rfl⟩⟩
+    have hpos :=
+      (pow_entry_pos_iff_exists_path (A := A) hA p.length i j).2 this
+    simpa using hpos
+  · intro h_exists
+    constructor
+    · exact hA
+    · intro i j
+      obtain ⟨k, hk_pos, hk_entry⟩ := h_exists i j
+      letI : Quiver n := toQuiver A
+      obtain ⟨⟨p, hp_len⟩⟩ :=
+        (pow_entry_pos_iff_exists_path (A := A) hA k i j).mp hk_entry
+      subst hp_len
+      exact ⟨p, hk_pos⟩
+
+/-- If a nonnegative square matrix `A` is primitive, then `A` is irreducible. -/
+theorem IsPrimitive.to_IsIrreducible
+    [IsOrderedRing R] [PosMulStrictMono R] [Nontrivial R] [DecidableEq n]
+    (h_prim : IsPrimitive A) : IsIrreducible A := by
+  obtain ⟨h_nonneg, k, hk_pos, hk_all⟩ := h_prim
+  refine ⟨h_nonneg, fun i j ↦ ?_⟩
+  letI : Quiver n := toQuiver A
+  have ⟨⟨p, hp⟩⟩ : Nonempty {p : Path i j // p.length = k} :=
+    (pow_entry_pos_iff_exists_path h_nonneg k i j).mp (hk_all i j)
+  exact ⟨p, hp ▸ hk_pos⟩
+
+/-! ## Transposition -/
+
+/-- Reverse a path in `toQuiver A` to a path in `toQuiver Aᵀ`, swapping endpoints. -/
+def transposeRev
+    {n R : Type*} [Ring R] [LinearOrder R] {A : Matrix n n R} :
+    ∀ {i j : n}, (@Quiver.Path n (toQuiver A) i j) →
+      (@Quiver.Path n (toQuiver Aᵀ) j i) := by
+  letI : Quiver n := toQuiver A
+  intro i j p
+  induction p with
+  | nil =>
+    exact (@Quiver.Path.nil _ (toQuiver Aᵀ) _)
+  | @cons b c q e ih =>
+    have eT : @Quiver.Hom n (toQuiver Aᵀ) c b := by
+      change 0 < (Aᵀ) c b
+      simpa [Matrix.transpose_apply] using e
+    exact (@Quiver.Path.comp n (toQuiver Aᵀ) c b i (@Quiver.Hom.toPath n (toQuiver Aᵀ) c b eT) ih)
+
+variable {n R : Type*} [Ring R] [LinearOrder R]
+variable {A : Matrix n n R}
+
+/-- Irreducibility is invariant under transpose. -/
+theorem IsIrreducible.transpose
+    (hA_nonneg : ∀ i j, 0 ≤ A i j) (hA : IsIrreducible A) : IsIrreducible Aᵀ := by
```

Comment:
The non-negativity assumption is already present in the irreducibility assumption:
```suggestion
theorem IsIrreducible.transpose (hA : IsIrreducible A) : IsIrreducible Aᵀ := by
```